src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java

```diff
@@ -820,4 +843,145 @@ public boolean hasAccessManagementPermission(final ApiKey apiKey) {
         return hasPermission(apiKey, Permissions.ACCESS_MANAGEMENT.name());
     }
 
+    public PaginatedResult getRootProjects(final boolean includeMetrics, final boolean excludeInactive) {
+        final PaginatedResult result;
+        final Query<Project> query = pm.newQuery(Project.class);
+        if (orderBy == null) {
+            query.setOrdering("name asc, version desc");
+        }
+
+        var filterBuilder = new ProjectQueryFilterBuilder()
+                .excludeInactive(excludeInactive)
+                .excludeChildren();
+
+        if (filter != null) {
+            final String filterString = ".*" + filter.toLowerCase() + ".*";
+            final Tag tag = getTagByName(filter.trim());
+
+            if (tag != null) {
+                filterBuilder = filterBuilder.withFuzzyNameOrExactTag(filterString, tag);
+
+            } else {
+                filterBuilder = filterBuilder.withFuzzyName(filterString);
+            }
+        }
+
+        final String queryFilter = filterBuilder.buildFilter();
+        final Map<String, Object> params = filterBuilder.getParams();
+
+        preprocessACLs(query, queryFilter, params, false);
+        pm.getFetchPlan().addGroup(Project.FetchGroup.ALL.name());
+        result = execute(query, params);
+        if (includeMetrics) {
+            // Populate each Project object in the paginated result with transitive related
+            // data to minimize the number of round trips a client needs to make, process, and render.
+            for (Project project : result.getList(Project.class)) {
+                project.setMetrics(getMostRecentProjectMetrics(project));
+            }
+        }
```

Comment:
Dieser Block scheint in sehr ähnlicher Form mehrfach vorzukommen in der Datei. Wenn da Teile gedoppelt sind, lager es in eine Funktion aus und ruf die in den jeweiligen Funktionen auf, um die Code-Dopplung zu umgehen

Reply:
Dadurch das die Root-Projects- und All-Parent-Relationships-Methode nun durch die schon vorhandenen Methoden mithilfe von zusätzlichen Query-Parametern ersetzt werden konnten, gibt es nur noch eine Code-Dopplung zwischen der Methode `getChildrenProjects` und den schon vorhandenen `getProjects`-Methoden. Allerdings würde ich diese nicht auslagern, da doppelt vorkommender Code in den originalen Methoden auch nicht ausgelagert wurde.

Screenshot aus `ProjectQueryManager.java` nach dem Entfernen von `getRootProjects` und `getProjectsWithParents`:

![grafik](https://user-images.githubusercontent.com/113189967/193227162-d991432e-c7a8-4627-addc-8046336857cd.png)